### PR TITLE
TST: Add test for col names during groupby().agg()

### LINKED ISSUE #42332: BUG: groupby().agg() loses column names for an empty dataframe with 'idxmax' as an aggregation function
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd

df = pd.DataFrame(columns=["id1", "id2", "time", "values"], dtype="int")


df_ok = df.groupby(["id1", "id2"]).sum()

print("Dataframe using .count() with groupby column names retained")
print(df_ok.index.names)


df_ok_2 = df.groupby(["id1", "id2"]).agg(**{
    "start": pd.NamedAgg(column="time", aggfunc="min"),
})

print()
print("Dataframe using aggfunc='min' with groupby columns retained")
print(df_ok_2.index.names)


df_bad = df.groupby(["id1", "id2"]).agg(**{
    "start": pd.NamedAgg(column="time", aggfunc="min"),
    "peak_time": pd.NamedAgg(column="values", aggfunc="idxmax")
})

print()
print("Dataframe using aggfunc='min' and aggfunc='idxmax' with groupby column names lost")
print(df_bad.index.names)


df_bad = df.groupby(["id1", "id2"]).agg(**{
    "peak_time": pd.NamedAgg(column="values", aggfunc="idxmax")
})

print()
print("Dataframe using only aggfunc='idxmax' with groupby columns lost entirely")
print(df_bad.index.names)
```

Output:
```
Dataframe using .count() with groupby column names retained
['id1', 'id2']

Dataframe using aggfunc='min' with groupby columns retained
['id1', 'id2']

Dataframe using aggfunc='min' and aggfunc='idxmax' with groupby column names lost
[None, None]

Dataframe using only aggfunc='idxmax' with groupby columns lost entirely
[None]
```

#### Problem description

The index returned by `DataFrame.groupby().agg()` is inconsistent depending on the aggregation applied to the groupby for an empty dataframe. This makes it difficult to use the output of these functions without adding subsequent if/else blocks to re-build the index names.

In the above examples, the provided DataFrame is empty. The behaviour of the returned DataFrames differs by how I handle the groupby:
1. If I use a function like .count(), the 'groupby' columns are both retained, and their names are kept, as I expect
2. If I use .agg() with 'min' as the function, the same good behaviour occurs
3. If I use .agg() with two functions, one being 'min' and the other 'idxmin', two index columns are returned but their names are lost, becoming 'level_0' and 'level_1'
4. If I use .agg() with only 'idxmin' as the function, only a single index column is returned, even though I grouped by two columns.

In Pandas 1.1.5, the third case behaved well, remembering the column names; the behaviour changed in 1.2.0.

#### Expected Output

The groupby columns and their names are remembered in all cases.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 7c48ff4409c622c582c56a5702373f726de08e96
python           : 3.8.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 3.10.0-1160.31.1.el7.x86_64
Version          : #1 SMP Tue Jun 15 10:20:52 CDT 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.2.5
numpy            : 1.19.4
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.3
setuptools       : 41.6.0
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : 3.5.4
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.1 (dt dec pq3 ext lo64)
jinja2           : 3.0.1
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 2021.06.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.2
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.7.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by calvh**
Column names should consistently be retained when using df.groupby().agg()

- [x] closes #42332
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Whats new:
Add a test to `pandas/tests/groupby/aggregate/test_aggregate.py`

The issue appears to be fixed now:

Tested on 1.4.0.dev0+508.g5fe02971f8

Screenshot from 2021-08-27 00-06-24

Output of `pd.show_versions()`

INSTALLED VERSIONS
------------------
commit           : 5fe02971f8bc40a6bbb742f1f376b3650c3c8eaa
python           : 3.9.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.11.0-31-generic
Version          : #33-Ubuntu SMP Wed Aug 11 13:19:04 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_CA.UTF-8
LOCALE           : en_CA.UTF-8

pandas           : 1.4.0.dev0+508.g5fe02971f8
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 20.3.4
setuptools       : 57.4.0
Cython           : 0.29.24
pytest           : 6.2.4
hypothesis       : 6.14.9
sphinx           : 4.1.2
blosc            : 1.10.4
feather          : None
xlsxwriter       : 3.0.1
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.26.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 2021.05.0
fastparquet      : 0.7.1
gcsfs            : 2021.05.0
matplotlib       : 3.4.3
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : 5.0.0
pyxlsb           : None
s3fs             : 2021.05.0
scipy            : 1.7.1
sqlalchemy       : 1.4.23
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.54.0

### CODE REVIEW

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 1]
```diff
@@ -1274,3 +1274,35 @@ def func(ser):
...
+
+
+def test_groupby_agg_column_names():
+    # GH42332
+
+    df = DataFrame(columns=["id1", "id2", "time", "values"], dtype="int").groupby(
+        ["id1", "id2"]
+    )
+
+    df_sum_idx = df.sum().index.names
```
[the remark this answers is not in the document]
Comment (calvh) [reply pandas/tests/groupby/aggregate/test_aggregate.py:1286]: @phofl sorry for the delay

After some experimenting, here is what I came up with to parametrize the test. I'm unsure of how to parametrize the `sum` function.

Also, I am sorry but could you elaborate on the meaning of the "whole dataframe" as I am not sure what you meant.

Thank you.

```python
@pytest.mark.parametrize(
    "agg_params",
    [
            {"start": pd.NamedAgg(column="time", aggfunc="min")},
            {
                "start": pd.NamedAgg(column="time", aggfunc="min"),
                "peak_time": pd.NamedAgg(column="values", aggfunc="idxmax"),
            },
            {"peak_time": pd.NamedAgg(column="values", aggfunc="idxmax")},
    ],
)
def test_groupby_agg_column_names(agg_params):
    # GH42332
    grouped = (
        DataFrame(columns=["id1", "id2", "time", "values"], dtype="int")
        .groupby(["id1", "id2"])
    )

    aggregated = grouped.agg(**agg_params)

    assert (
        grouped.sum().index.names == aggregated.index.names == ["id1", "id2"]
    )

```

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 2]
```diff
@@ -1274,3 +1274,35 @@ def func(ser):
...
+        }
+    ).index.names
+
+    df_agg3_idx = df.agg(
+        **{"peak_time": pd.NamedAgg(column="values", aggfunc="idxmax")}
+    ).index.names
+
+    expected = ["id1", "id2"]
+
+    assert df_sum_idx == expected
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:1305]: use `tm.assert_frame_equal` and construct the actual expected value.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by calvh** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.

Sorry, yes I am still interested in working on this

**Comment by mroeschke** [comment]
Appears this PR has been dormant for a while and still needs updates so closing. If interested in continuing, please merge master, address related comments and we can reopen.

**Pull request closed without merging**